File: tests/passjoin-index.js
'use strict';

const standardLevenshtein = require('leven');
const protocol = process.env.MNEMONIST_TRANSPORT === 'protocol'
  ? require('../adapter/_protocol.js')
  : null;

function comparator(a, b) {
  return a.length > b.length ? -1 : a.length < b.length ? 1 : a < b ? -1 : a > b ? 1 : 0;
}

function partition(k, length) {
  const segments = k + 1;
  const small = length / segments | 0;
  const large = small + 1;
  const largeCount = length - small * segments;
  const smallCount = segments - largeCount;
  const result = new Array(segments);
  let i = 0;
  for (; i < smallCount; i++) result[i] = [i * small, small];
  const offset = i * small;
  for (let j = 0; j < largeCount; j++) result[i + j] = [offset + j * large, large];
  return result;
}

function segments(k, string) {
  return partition(k, string.length).map(part => string.slice(part[0], part[0] + part[1]));
}

function segmentPos(k, i, string) {
  return partition(k, string.length)[i][0];
}

function multiMatchAwareInterval(k, delta, i, s, pi, li) {
  const start1 = pi - i;
  const end1 = pi + i;
  const remaining = k - i;
  const start2 = pi + delta - remaining;
  const end2 = pi + delta + remaining;
  return [Math.max(0, start1, start2), Math.min(end1, end2, s - li)];
}

function multiMatchAwareSubstrings(k, string, length, i, pi, li) {
  const interval = multiMatchAwareInterval(k, string.length - length, i, string.length, pi, li);
  const result = [];
  let previous = null;
  for (let start = interval[0]; start <= interval[1]; start++) {
    const substring = string.slice(start, start + li);
    if (substring !== previous) result.push(substring);
    previous = substring;
  }
  return result;
}

function PassjoinIndex(levenshtein, k) {
  if (typeof levenshtein !== 'function') {
    throw new Error('mnemonist/passjoin-index: `levenshtein` should be a function returning edit distance between two strings.');
  }
  if (typeof k !== 'number' || k < 1) throw new Error('mnemonist/passjoin-index: `k` should be a number > 0');
  this.levenshtein = levenshtein;
  this.k = k;
  this._inner = protocol && levenshtein === standardLevenshtein
    ? protocol.create('passjoin-index', [k])
    : null;
  this.clear();
}

PassjoinIndex.prototype.clear = function() {
  this.size = 0;
  this.strings = [];
  if (this._inner) this._inner.call('clear', []);
};
PassjoinIndex.prototype.add = function(value) {
  if (this._inner) {
    this._inner.call('add', [value]);
    this.size = this._inner.size();
    return this;
  }
  this.strings.push(value);
  this.size++;
  return this;
};
PassjoinIndex.prototype.search = function(query) {
  if (this._inner) return new Set(this._inner.call('search', [query]));
  const matches = new Set();
  for (const candidate of this.strings) {
    if (Math.abs(candidate.length - query.length) <= this.k && this.levenshtein(query, candidate) <= this.k) matches.add(candidate);
  }
  return matches;
};
PassjoinIndex.prototype.forEach = function(callback, scope) {
  const context = arguments.length > 1 ? scope : this;
  if (this._inner) {
    this._inner.call('values', []).forEach((value, index) => callback.call(context, value, index, this));
    return;
  }
  this.strings.forEach((value, index) => callback.call(context, value, index, this));
};
PassjoinIndex.prototype.values = function() {
  return this._inner ? this._inner.call('values', []).values() : this.strings.values();
};
PassjoinIndex.prototype[Symbol.iterator] = PassjoinIndex.prototype.values;
PassjoinIndex.from = function(iterable, levenshtein, k) {
  const index = new PassjoinIndex(levenshtein, k);
  for (const value of iterable) index.add(value);
  return index;
};

PassjoinIndex.comparator = protocol
  ? (a, b) => protocol.invoke('passjoin-index', 'comparator', [a, b])
  : comparator;
PassjoinIndex.partition = protocol
  ? (k, length) => protocol.invoke('passjoin-index', 'partition', [k, length])
  : partition;
PassjoinIndex.segments = protocol
  ? (k, string) => protocol.invoke('passjoin-index', 'segments', [k, string])
  : segments;
PassjoinIndex.segmentPos = protocol
  ? (k, i, string) => protocol.invoke('passjoin-index', 'segmentPos', [k, i, string])
  : segmentPos;
PassjoinIndex.multiMatchAwareInterval = protocol
  ? (k, delta, i, s, pi, li) => protocol.invoke('passjoin-index', 'multiMatchAwareInterval', [k, delta, i, s, pi, li])
  : multiMatchAwareInterval;
PassjoinIndex.multiMatchAwareSubstrings = protocol
  ? (k, string, length, i, pi, li) => protocol.invoke('passjoin-index', 'multiMatchAwareSubstrings', [k, string, length, i, pi, li])
  : multiMatchAwareSubstrings;

module.exports = PassjoinIndex;

File: src/passjoin_index.rs
use indexmap::IndexMap;
use serde_json::Value;

use crate::bk_tree::levenshtein;

pub fn comparator(left: &str, right: &str) -> std::cmp::Ordering {
    right
        .chars()
        .count()
        .cmp(&left.chars().count())
        .then_with(|| left.cmp(right))
}

pub fn partition(k: usize, length: usize) -> Vec<(usize, usize)> {
    let segments = k + 1;
    let small = length / segments;
    let large = small + 1;
    let large_count = length - small * segments;
    let small_count = segments - large_count;
    let mut result = Vec::with_capacity(segments);

    for index in 0..small_count {
        result.push((index * small, small));
    }
    let offset = small_count * small;
    for index in 0..large_count {
        result.push((offset + index * large, large));
    }
    result
}

pub fn segments(k: usize, value: &str) -> Vec<String> {
    let characters: Vec<_> = value.chars().collect();
    partition(k, characters.len())
        .into_iter()
        .map(|(start, length)| characters[start..start + length].iter().collect())
        .collect()
}

pub fn segment_pos(k: usize, index: usize, value: &str) -> usize {
    partition(k, value.chars().count())[index].0
}

pub fn multi_match_aware_interval(
    k: isize,
    delta: isize,
    index: isize,
    string_length: isize,
    position: isize,
    segment_length: isize,
) -> (isize, isize) {
    let start = (position - index)
        .max(position + delta - (k - index))
        .max(0);
    let end = (position + index)
        .min(position + delta + (k - index))
        .min(string_length - segment_length);
    (start, end)
}

pub fn multi_match_aware_substrings(
    k: isize,
    value: &str,
    target_length: isize,
    index: isize,
    position: isize,
    segment_length: isize,
) -> Vec<String> {
    let characters: Vec<_> = value.chars().collect();
    let (start, end) = multi_match_aware_interval(
        k,
        characters.len() as isize - target_length,
        index,
        characters.len() as isize,
        position,
        segment_length,
    );
    let mut result = Vec::new();
    let mut previous = None;
    for offset in start..=end {
        let substring: String = characters[offset as usize..(offset + segment_length) as usize]
            .iter()
            .collect();
        if previous.as_ref() != Some(&substring) {
            previous = Some(substring.clone());
            result.push(substring);
        }
    }
    result
}

#[derive(Debug, Clone, Default)]
pub struct PassjoinIndex {
    records: IndexMap<String, Value>,
}

impl PassjoinIndex {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn size(&self) -> usize {
        self.records.len()
    }

    pub fn clear(&mut self) {
        self.records.clear();
    }

    pub fn add(&mut self, key: impl Into<String>, value: Value) {
        self.records.insert(key.into(), value);
    }

    pub fn search(&self, query: &str, radius: usize) -> Vec<(&String, &Value, usize)> {
        let mut results: Vec<_> = self
            .records
            .iter()
            .filter_map(|(key, value)| {
                let distance = levenshtein(key, query);
                (distance <= radius).then_some((key, value, distance))
            })
            .collect();
        results.sort_by(|a, b| a.2.cmp(&b.2).then_with(|| a.0.cmp(b.0)));
        results
    }
}
